File: leetcode/138_Copy_List_with_Random_Pointer/solution.cpp
/* 
 * 138 Copy List with Random Pointer 
 *
 * A linked list is given such that each node contains an additional random pointer which could point to any node in the list or null.
 * Return a deep copy of the list.
 *
 * */

#include <iostream>
#include <assert.h>

/*
 * Definition for singly-linked list with a random pointer.
 */ 
struct RandomListNode {
	int label;
 	RandomListNode *next, *random;
    RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
};

class Solution {
	public:
		RandomListNode *copyRandomList(RandomListNode *head) {
			if (head == NULL) {
				return NULL;
			}

			RandomListNode *h = head;
			RandomListNode *node;
			RandomListNode *temp;
			RandomListNode dummy(0);

			while(h) {
				node = new RandomListNode(h->label);
				node->random = h->random;
				temp = h->next;
				h->next = node;
				node->next = temp;
				h = temp;
			}

			std::cout << "step 0" << std::endl;
			h = head;
			while(h != NULL) {
				std::cout << h->label << std::endl;	
				h = h->next;
			}

			std::cout << "step 1" << std::endl;
			h = head->next;
			while(h) {
				if (h->random != NULL) {
					h->random = h->random->next;				
				}
				if (h->next)
					h = h->next->next;
				else 
					break;
			}

			std::cout << "step 2" << std::endl;
			h = head;
			node = &dummy;
			while(h) {
				node->next = h->next;
				temp = h->next->next;
				h->next = temp;
				h = h->next;
				node = node->next;
			}
			
			std::cout << "step 3" << std::endl;

			return dummy.next;
		}

		void printList(RandomListNode* head) {
			if(head == NULL) {
				return ;
			} 

			std::cout << head->label << std::endl;	
			while(head->next != NULL) {
				head = head->next;
				std::cout << head->label << std::endl;	
			}
		}
	
		void freeList(RandomListNode* head) {
			if(head == NULL) {
				return ;
			} 

			RandomListNode* tmp = head->next;
			while(head != NULL) {
				tmp = head->next;
				delete head;
				head = tmp;	
			}
		}

};

int  main() {
	/*RandomListNode *a3 = new RandomListNode(4);
	RandomListNode *a2 = new RandomListNode(3);
	a2->next = a3;
	*/
	RandomListNode *a1 = new RandomListNode(-1);
	//a1->next = a2;
	RandomListNode *a = new RandomListNode(-1);
	a->next = a1;
	//a->random = a3;

	Solution *test = new Solution();
	test->printList(a);
	RandomListNode *head = test->copyRandomList(a);	
	std::cout << "finish." << std::endl;
	test->printList(head);
	test->freeList(head);

	return 0;	
}

File: leetcode/138_Copy_List_with_Random_Pointer/Solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

class Solution {
    public static class Node {
        int val;
        Node next;
        Node random;

        public Node(int val) {
            this.val = val;
            this.next = null;
            this.random = null;
        }
    }

    public Node copyRandomList(Node head) {

        if (head == null)
            return null;

        Node cur = head;
        ArrayList<Node> tmpList = new ArrayList<Node>();
        Map<Node, Integer> tmpMap = new HashMap<Node, Integer>();

        int index = 0;
        while (cur != null) {
            Node newNode = new Node(cur.val);
            tmpList.add(newNode);

            tmpMap.put(cur, index); 
            ++index;

            cur = cur.next;
        }

        for (int i = 0; i < tmpList.size() - 1; ++i) {
            tmpList.get(i).next = tmpList.get(i+1);
        }
        
        cur = head;
        int i = 0;
        while (cur != null) {

            if (cur.random == null) {
                tmpList.get(i).random = null;
            } else {
                int randIndex = tmpMap.get(cur.random);
                tmpList.get(i).random = tmpList.get(randIndex); 
            }
            cur = cur.next;
            ++i;
        }

        return tmpList.get(0);
    }

    public static void main(String[] args) {
        int target = 3;
        Node a1 = new Node(7);
        Node a2 = new Node(13);
        Node a3 = new Node(11);
        Node a4 = new Node(10);
        Node a5 = new Node(1);

        a1.next = a2;
        a2.next = a3;
        a3.next = a4;
        a4.next = a5;

        a2.random = a1;
        a3.random = a5;
        a4.random = a3;
        a5.random = a1;

        Solution test = new Solution();
        test.copyRandomList(a1);
    }

}